In [2]:
#2. Relier les bulles connectées entre elles 

import dash
from dash import html, dcc, Input, Output
import plotly.graph_objects as go
import networkx as nx
import pandas as pd



### Etude du fichier csv et compte des étiquettes 

df = pd.read_csv("Data_20250506.csv", sep=';')

etiquettes_data = df.iloc[:, 11:]  # à partir de la colonne 12, on prend tous les mots clés des études scientifiques 
etiquettes_series = pd.Series(etiquettes_data.values.ravel())

# Nettoyage des valeurs : on enlève les cases vides, les majuscules...car sinon on compterait différemment Fer et fer par exemple

etiquettes_series = etiquettes_series.dropna().astype(str).str.strip().str.lower() 

etiquettes_comptes = etiquettes_series.value_counts().reset_index()  # compte le nombre d'occurences dans toute la littérature scientifique
etiquettes_comptes.columns = ['Étiquette', 'Nombre_de_récurrence']

print(etiquettes_comptes)



#### A ce point la, certains mots sont encore compté deux fois différement  = electric vehicles et elecric vehicle par exmple !



                                Étiquette  Nombre_de_récurrence
0                                       0                 67495
1                       electric vehicles                  2675
2                         electrification                  1253
3                        electric vehicle                   558
4               battery electric vehicles                   437
...                                   ...                   ...
23195                  system variability                     1
23196         generation plant investment                     1
23197                  total system costs                     1
23198  variable renewable energy sources.                     1
23199                        resource use                     1

[23200 rows x 2 columns]


In [3]:
#tokenise 
import nltk
from nltk.tokenize import word_tokenize

nltk.download('punkt')
nltk.download('punkt_tab')




c:\Users\ewenm\miniconda3\Lib\site-packages\nltk\metrics\association.py:26: UserWarning: A NumPy version >=1.22.4 and <2.3.0 is required for this version of SciPy (detected version 2.3.1)
  from scipy.stats import fisher_exact
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\ewenm\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     C:\Users\ewenm\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


True

In [4]:


text=df['AB_AI'][0]
print(text)

tokens = word_tokenize(text)

Electric vehicles have been developed with the aims of saving energy and reducing carbon dioxide emissions. Furthermore, as a single bus can transport many people, buses produce low carbon dioxide emissions per person in comparison with cars. In recent years, community buses have been introduced to ensure the mobility of senior citizens. Therefore, a low-floor vehicle was adopted as the base vehicle in our latest project. The bus was demonstrated around the center of the city. It followed a circular route of distance 7.2 kilometers, the number of bus stops on the route was 27 and the driving time was 40 minutes per circuit. This demonstration was repeated three to four times a day for 14 days. During the 86 circuits of the demonstration, a total of 1,110 passengers used the bus and the total running distance was 776 kilometers. From these results, we concluded that the quantity of carbon dioxide emissions could be reduced by 37% using our proposed bus in comparison with that of a conve

In [5]:
#Virons les stop words 

from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.corpus import stopwords
import nltk

nltk.download('stopwords')

stopwords = set(stopwords.words('english'))
print('nombre de mots de text :' , len(text))
words = word_tokenize(text)
words_filtered = []
print('nombre de mots de words :' , len(words))
# on va renvoyer une liste des mots en ne gardant "que les plus importants". On enlève par exemple les "is", "of... En somme, on ne garde que les mots clés.
for w in words:
  if w not in stopwords:
    words_filtered.append(w)



# J'enlève maintenant toute la ponctuation de ma liste de mots. 
ponctuation = ['.',',',';','?',':','!','%']
words_sans_ponctuation=[x for x in words_filtered if x not in ponctuation]


# on a beaucoup de doublons dans la liste : on les retire pour diminuer le nombre de mots 
W=list(set(words_sans_ponctuation))
print('nombre de mots de W ( sans ponctuation) :' , len(W))


# on fait un choix : on enlève les mots qui terminent par "ED" en anglais car on estime que ce sont des verbes, et non les principaux mots-clés. 
# parce que les encadrants préfèrent. On pourrait enlever cette ligne. 

W_sans_ED = [w for w in W if not w.lower().endswith("ed")] 
print('nombre de mots de W_sans_ED :' , len(W_sans_ED))
print(W_sans_ED)

# A ce stade on a réussi à réduire le nombre de mots clés de 1026 mots à 79. Ca rste encore beaucoup trop si on veut traiter beaucoup d'articles en meme temps. 
sans_maj = [w.lower() for w in W_sans_ED]

nombre de mots de text : 1026
nombre de mots de words : 187
nombre de mots de W ( sans ponctuation) : 88
nombre de mots de W_sans_ED : 78
['people', 'saving', 'emissions', 'could', 'During', 'low-floor', 'The', '776', 'time', 'mobility', 'city', 'aims', 'produce', 'running', 'years', 'results', 'number', 'four', 'base', 'per', '14', 'conventional', 'quantity', 'stops', 'distance', 'reducing', 'times', 'transport', 'In', 'Furthermore', 'From', 'three', 'driving', 'days', 'day', '27', 'Therefore', 'bus', 'center', 'single', 'using', 'Electric', 'cars', '37', 'circuit', 'circular', 'recent', 'total', 'person', 'It', 'carbon', 'vehicle', 'kilometers', 'demonstration', 'route', 'low', 'comparison', 'citizens', 'This', 'ensure', 'senior', 'vehicles', '7.2', '86', 'diesel-engine', 'around', 'many', 'passengers', 'dioxide', 'buses', 'circuits', 'community', '1,110', 'energy', 'latest', 'minutes', 'project', '40']


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\ewenm\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [6]:
# on enlève les majuscules pour pouvoir enlever les stopwords qui contenaient des majuscules 
mots_filtrés =[]
for w in sans_maj:
  if w not in stopwords:
    mots_filtrés.append(w)

print('nombre de mots de mots_filtrés :' , len(mots_filtrés))
print(mots_filtrés)

nombre de mots de mots_filtrés : 72
['people', 'saving', 'emissions', 'could', 'low-floor', '776', 'time', 'mobility', 'city', 'aims', 'produce', 'running', 'years', 'results', 'number', 'four', 'base', 'per', '14', 'conventional', 'quantity', 'stops', 'distance', 'reducing', 'times', 'transport', 'furthermore', 'three', 'driving', 'days', 'day', '27', 'therefore', 'bus', 'center', 'single', 'using', 'electric', 'cars', '37', 'circuit', 'circular', 'recent', 'total', 'person', 'carbon', 'vehicle', 'kilometers', 'demonstration', 'route', 'low', 'comparison', 'citizens', 'ensure', 'senior', 'vehicles', '7.2', '86', 'diesel-engine', 'around', 'many', 'passengers', 'dioxide', 'buses', 'circuits', 'community', '1,110', 'energy', 'latest', 'minutes', 'project', '40']


In [7]:
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet

nltk.download('wordnet')
nltk.download('omw-1.4')

lemmatizer = WordNetLemmatizer()

# Mise au singulier
W_singuliers = list(set([lemmatizer.lemmatize(w, pos='n') for w in mots_filtrés]))

print('nombre de mots de W_singuliers', len(W_singuliers))
print(W_singuliers)

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\ewenm\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\ewenm\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


nombre de mots de W_singuliers 67
['four', 'base', 'per', '14', 'conventional', 'people', 'quantity', 'result', 'therefore', 'ensure', 'senior', 'saving', 'passenger', 'distance', '37', 'could', '7.2', '86', 'diesel-engine', 'aim', 'circuit', 'around', 'reducing', 'low-floor', 'minute', 'circular', 'year', 'many', 'recent', 'furthermore', 'total', 'dioxide', 'person', '776', 'car', 'transport', 'electric', 'time', 'carbon', 'emission', 'mobility', 'vehicle', 'city', 'three', 'produce', 'kilometer', 'community', 'driving', 'stop', 'running', 'demonstration', '1,110', 'energy', 'day', 'latest', '27', 'route', 'citizen', 'low', 'project', 'bus', 'center', 'comparison', 'number', 'single', 'using', '40']


In [8]:
# j'enlève les mots de moins de 3 caractères ; 
mots_assez_long = []
for w in W_singuliers : 
    if len(w)>=3 : 
        mots_assez_long.append(w)

print('nombre de mots de mots_assez_long', len(mots_assez_long))
print(mots_assez_long)

nombre de mots de mots_assez_long 62
['four', 'base', 'per', 'conventional', 'people', 'quantity', 'result', 'therefore', 'ensure', 'senior', 'saving', 'passenger', 'distance', 'could', '7.2', 'diesel-engine', 'aim', 'circuit', 'around', 'reducing', 'low-floor', 'minute', 'circular', 'year', 'many', 'recent', 'furthermore', 'total', 'dioxide', 'person', '776', 'car', 'transport', 'electric', 'time', 'carbon', 'emission', 'mobility', 'vehicle', 'city', 'three', 'produce', 'kilometer', 'community', 'driving', 'stop', 'running', 'demonstration', '1,110', 'energy', 'day', 'latest', 'route', 'citizen', 'low', 'project', 'bus', 'center', 'comparison', 'number', 'single', 'using']


In [9]:
import re

def est_nombre(w):
    return re.fullmatch(r"[0-9]+([.,][0-9]+)?", w) is not None

W_sans_nombres = [w for w in mots_assez_long if not est_nombre(w)]

print('nombre de mots de W_sans_nombres', len(W_sans_nombres))
print(W_sans_nombres)


nombre de mots de W_sans_nombres 59
['four', 'base', 'per', 'conventional', 'people', 'quantity', 'result', 'therefore', 'ensure', 'senior', 'saving', 'passenger', 'distance', 'could', 'diesel-engine', 'aim', 'circuit', 'around', 'reducing', 'low-floor', 'minute', 'circular', 'year', 'many', 'recent', 'furthermore', 'total', 'dioxide', 'person', 'car', 'transport', 'electric', 'time', 'carbon', 'emission', 'mobility', 'vehicle', 'city', 'three', 'produce', 'kilometer', 'community', 'driving', 'stop', 'running', 'demonstration', 'energy', 'day', 'latest', 'route', 'citizen', 'low', 'project', 'bus', 'center', 'comparison', 'number', 'single', 'using']


On a réussi à passer de 1026 mots à 59 ! Il me faudrait maintenant une IA pour enlever les mots ayant moins d'intéret en fonction du sujet pour sélectionner les mots clé importants.

In [10]:

print(len(df['AB_AI']))




7100


In [11]:
#Faisons une fonction plus rapide pour sélectionner les mots importants d'un titre d'article : 


from nltk.corpus import stopwords
from nltk.tokenize import sent_tokenize, word_tokenize
from nltk.corpus import stopwords
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.corpus import wordnet
import re

nltk.download('stopwords')
stopwords = set(stopwords.words('english'))
nltk.download('wordnet')
nltk.download('omw-1.4')

def est_nombre(w):
        return re.fullmatch(r"[0-9]+([.,][0-9]+)?", w) is not None


def mots_cles(text) : 


    words = word_tokenize(text)
    sans_maj = [w.lower() for w in words]
    words_filtered = []
    for w in sans_maj:
      if w not in stopwords:
        words_filtered.append(w)


#ponctuation 
    ponctuation = ['.',',',';','?',':','!','%']
    words_sans_ponctuation=[x for x in words_filtered if x not in ponctuation]


# doublons
    W=list(set(words_sans_ponctuation))

# sans les -ED
    W_sans_ED = [w for w in W if not w.lower().endswith("ed")] 

# singulier

    lemmatizer = WordNetLemmatizer()
    W_singuliers = list(set([lemmatizer.lemmatize(w, pos='n') for w in W_sans_ED]))

#sans les mots très courts
    mots_assez_long = []
    for w in W_singuliers : 
        if len(w)>=3 : 
            mots_assez_long.append(w)

# sans les nombres 
   
    W_sans_nombres = [w for w in mots_assez_long if not est_nombre(w)]

    return(W_sans_nombres)





[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\ewenm\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\ewenm\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\ewenm\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


In [12]:
L = []

#On vai maintenant parcourir tous les titres !
for i in range (len(df['AB_AI'])) : 
    L.append(mots_cles(df['AB_AI'][i]))


liste = [mot for sous_liste in L for mot in sous_liste]
print('nombre de mots dans liste :',len(liste))  
phrase = ' '.join(liste)

print('nombre de mots dans phrase : ',len(mots_cles(phrase))) 

nombre de mots dans liste : 570379
nombre de mots dans phrase :  18409


In [13]:
from collections import Counter


# Je range dans un dictionnaire le mot et son nombre d'apparition dans l'ensemble des titres, et ce par ordre décroissant d'apparition.
compteur = Counter(liste)
dico_trie = dict(compteur.most_common())



In [14]:
# Je supprime tous les mots qui apparaissent moins de 500 fois en supposant ainsi que ce sont des mots peu importants, que je n'avais pas pu supprimer avant.

compteur_filtre = {mot: nb for mot, nb in compteur.items() if nb >= 500}

compteur_filtre_trie = dict(sorted(compteur_filtre.items(), key=lambda x: x[1], reverse=True))

print('On a réduit le nombre total de "mots-clés" de :',len(liste), "à ",len(compteur_filtre_trie))
print(compteur_filtre_trie)


On a réduit le nombre total de "mots-clés" de : 570379 à  234
{'electric': 5740, 'vehicle': 5629, 'result': 3733, 'energy': 3708, 'system': 3623, 'study': 3277, 'model': 3163, 'paper': 3098, 'battery': 2693, 'charging': 2587, 'power': 2513, 'transport': 2298, 'using': 2278, 'electrification': 2077, 'different': 2068, 'emission': 2042, 'cost': 2026, 'also': 1998, 'show': 1981, 'analysis': 1940, 'however': 1855, 'impact': 1845, 'demand': 1806, 'technology': 1711, 'method': 1706, 'strategy': 1635, 'time': 1633, 'potential': 1578, 'performance': 1578, 'approach': 1546, 'data': 1524, 'use': 1479, 'fuel': 1471, 'increase': 1459, 'simulation': 1411, 'due': 1408, 'electricity': 1405, 'present': 1402, 'current': 1389, 'significant': 1387, 'high': 1379, 'carbon': 1376, 'reduce': 1364, 'network': 1308, 'sector': 1301, 'new': 1298, 'transportation': 1293, 'case': 1292, 'two': 1289, 'research': 1284, 'solution': 1282, 'scenario': 1260, 'operation': 1244, 'consumption': 1242, 'one': 1242, 'optimizat

## En pure comparaison, j'ai demandé à chatGPT de me choisir les mots importants dans la liste L avec les 570000 mots : voici le résultat : 

In [15]:
dico_filtré = {
    'electric': 5740,
    'vehicle': 5629,
    'energy': 3708,
    'system': 3624,
    'study': 3278,
    'model': 3164,
    'battery': 2693,
    'charging': 2587,
    'power': 2513,
    'transport': 2298,
    'electrification': 2077,
    'emission': 2042,
    'cost': 2026,
    'analysis': 1940,
    'impact': 1845,
    'demand': 1806,
    'technology': 1711,
    'method': 1708,
    'strategy': 1635,
    'performance': 1578,
    'approach': 1546,
    'data': 1524,
    'fuel': 1471,
    'simulation': 1411,
    'electricity': 1405,
    'carbon': 1376,
    'reduce': 1364,
    'network': 1310,
    'sector': 1301,
    'transportation': 1294,
    'research': 1285,
    'solution': 1282,
    'operation': 1244,
    'consumption': 1242,
    'optimization': 1241,
    'grid': 1230,
    'efficiency': 1225,
    'infrastructure': 1205,
    'development': 1172,
    'environmental': 1171,
    'management': 1152,
    'reduction': 1133,
    'optimal': 1131,
    'range': 1126,
    'hybrid': 1119,
    'renewable': 1110,
    'policy': 1100,
    'driving': 1091,
    'source': 1069,
    'algorithm': 1069,
    'station': 1043,
    'capacity': 1020,
    'design': 978,
    'storage': 978,
    'distribution': 973,
    'process': 940,
    'dynamic': 934,
    'application': 917,
    'generation': 874,
    'charge': 871,
    'economic': 863,
    'adoption': 846,
    'framework': 844,
    'conventional': 827,
    'fleet': 819,
    'public': 818,
    'urban': 812,
    'dioxide': 809,
    'greenhouse': 801,
    'sustainable': 792,
    'parameter': 781,
    'transition': 760,
    'behavior': 760,
    'key': 759,
    'mobility': 725,
    'cycle': 721,
    'supply': 712,
    'characteristic': 704,
    'service': 699,
    'reducing': 679,
    'engine': 677,
    'operating': 664,
    'climate': 654,
    'cell': 635,
    'alternative': 633,
    'planning': 630,
    'internal': 628,
    'modeling': 626,
    'efficient': 621,
    'mode': 619,
    'industry': 618,
    'air': 615,
    'objective': 609,
    'combustion': 608,
    'environment': 605,
    'production': 602,
    'effective': 591,
    'integration': 559,
    'electrical': 545,
    'bus': 538,
    'smart': 502,
}


In [16]:
print(len(dico_filtré)) 

101


On passe de 234 mots avec ma méthode à 101 mots avec chatGPT. Sur 570 000 mots d'origine, je considère que ma méthode est plutôt satisfaisante, surtout que les 101 mots sont bien compris 

In [17]:
# On veut maintenant calculer des poids d'importance des mots pour qu'on puisse attribuer à chaque mot un rayon plus ou moins grand. 
# On fait une echelle pour avoir des rayons compris entre 10 et 100. 
def scale(freq, min_freq, max_freq, min_scale=10, max_scale=100):
    return min_scale + (freq - min_freq) * (max_scale - min_scale) / (max_freq - min_freq)

min_freq = min(dico_filtré.values())
max_freq = max(dico_filtré.values())

rayon_mots = {
    mot: round(scale(freq, min_freq, max_freq), 2)
    for mot, freq in dico_filtré.items()
}

print(rayon_mots)

{'electric': 100.0, 'vehicle': 98.09, 'energy': 65.09, 'system': 63.64, 'study': 57.7, 'model': 55.74, 'battery': 47.65, 'charging': 45.82, 'power': 44.55, 'transport': 40.86, 'electrification': 37.06, 'emission': 36.46, 'cost': 36.19, 'analysis': 34.71, 'impact': 33.08, 'demand': 32.41, 'technology': 30.77, 'method': 30.72, 'strategy': 29.47, 'performance': 28.49, 'approach': 27.94, 'data': 27.56, 'fuel': 26.65, 'simulation': 25.62, 'electricity': 25.52, 'carbon': 25.02, 'reduce': 24.81, 'network': 23.88, 'sector': 23.73, 'transportation': 23.61, 'research': 23.45, 'solution': 23.4, 'operation': 22.75, 'consumption': 22.71, 'optimization': 22.7, 'grid': 22.51, 'efficiency': 22.42, 'infrastructure': 22.08, 'development': 21.51, 'environmental': 21.49, 'management': 21.17, 'reduction': 20.84, 'optimal': 20.81, 'range': 20.72, 'hybrid': 20.6, 'renewable': 20.45, 'policy': 20.27, 'driving': 20.12, 'source': 19.74, 'algorithm': 19.74, 'station': 19.3, 'capacity': 18.9, 'design': 18.18, 'st

In [18]:
from collections import defaultdict
import itertools
import pandas as pd


tous_mots = list(dico_filtré)

cooc_matrice = defaultdict(lambda: defaultdict(int))

# On parcourt chaque sous-liste de mots (chaque titre)
for mots in L:
    mots_filtres = [m for m in set(mots) if m in tous_mots] 
    for w1, w2 in itertools.combinations(mots_filtres, 2):
        cooc_matrice[w1][w2] += 1
        cooc_matrice[w2][w1] += 1  #comme on aura une matrice symétrique, on essaie de réduire la complexité en ne parcourant pas deux fois
        
df_cooc = pd.DataFrame(cooc_matrice, index=tous_mots, columns=tous_mots).fillna(0).astype(int)


print(df_cooc.shape)  
print(df_cooc.head())


(101, 101)
          electric  vehicle  energy  system  study  model  battery  charging  \
electric         0     5249    2970    2936   2721   2667     2454      2469   
vehicle       5249        0    2916    2853   2653   2612     2359      2355   
energy        2970     2916       0    2220   1699   1672     1551      1346   
system        2936     2853    2220       0   1660   1661     1448      1383   
study         2721     2653    1699    1660      0   1552     1265      1253   

          power  transport  ...  air  objective  combustion  environment  \
electric   2124       1555  ...  486        534         542          493   
vehicle    2022       1465  ...  490        521         555          485   
energy     1669       1312  ...  333        378         321          352   
system     1679       1191  ...  324        356         226          326   
study      1078       1059  ...  281        302         302          280   

          production  effective  integration  elect

In [19]:
### Liens entre les bulles de mots 

liens= []

for a, b in itertools.combinations(df_cooc.columns, 2):
    if df_cooc.loc[a, b] > 1000:  # on choisi de garder que les mots que reviennent ensemble de les titres plus de 1000 fois.
        liens.append((a, b))

print(f"{len(liens)} liens trouvés avec cooccurrence > 1000")
print(liens)  # aperçu des premières paires



116 liens trouvés avec cooccurrence > 1000
[('electric', 'vehicle'), ('electric', 'energy'), ('electric', 'system'), ('electric', 'study'), ('electric', 'model'), ('electric', 'battery'), ('electric', 'charging'), ('electric', 'power'), ('electric', 'transport'), ('electric', 'electrification'), ('electric', 'emission'), ('electric', 'cost'), ('electric', 'analysis'), ('electric', 'impact'), ('electric', 'demand'), ('electric', 'technology'), ('electric', 'method'), ('electric', 'strategy'), ('electric', 'performance'), ('electric', 'approach'), ('electric', 'data'), ('electric', 'fuel'), ('electric', 'simulation'), ('electric', 'electricity'), ('electric', 'carbon'), ('electric', 'reduce'), ('electric', 'network'), ('electric', 'transportation'), ('electric', 'research'), ('electric', 'solution'), ('electric', 'operation'), ('electric', 'consumption'), ('electric', 'optimization'), ('electric', 'grid'), ('electric', 'infrastructure'), ('electric', 'management'), ('electric', 'optimal'

Idées à ce stade : enlever tous les mots qui n'ont pas de liens avec les autres parces qu'ils servent à rien en gros.


Ne faire apparaitre les lies de chaque mot que lorsqu'on clique dessus


In [20]:
import http

http -F https://Bob:hiccup@ollama-sam.inria.fr/api/tags | jq '.models.[].name'

SyntaxError: invalid syntax (1938022865.py, line 3)

In [ ]:
import requests



In [ ]:
request = requests.get("http://ollama.pl.sophia.inria.fr:8080/api/tags")

ConnectionError: HTTPConnectionPool(host='ollama.pl.sophia.inria.fr', port=8080): Max retries exceeded with url: /api/tags (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001FA8D642300>: Failed to establish a new connection: [WinError 10061] Aucune connexion n’a pu être établie car l’ordinateur cible l’a expressément refusée'))

In [ ]:
request.status_code

200

In [ ]:
for n in request.json()['models'] : 
    print(n['name'])

qwen2.5-coder:32b
qwen2.5-coder:14b
qwen2.5-coder:7b
codellama:13b
starcoder2:latest
vicuna:latest
starcoder2:instruct
sqlcoder:7b
starcoder2:15b
qwen:7b
qwen:32b
qwen2.5:32b
openchat:7b
nous-hermes2-mixtral:8x7b
nomic-embed-text:137m-v1.5-fp16
moondream:latest
mxbai-embed-large:335m-v1-fp16
mistral:7b
mixtral:8x7b
mistral-large:123b
mixtral:8x22b
llava:34b-v1.6
llava:v1.6
llava:13b-v1.6
llama3:8b
llama3.1:8b
llama2:13b
llama2:latest
llama2:70b
gemma:7b-instruct
gemma2:9b
gemma2:27b
gemma2:2b
falcon:7b-instruct
deepseek-coder:6.7b
deepseek-coder-v2:16b
falcon:40b
deepseek-coder:1.3b
codestral:22b
all-minilm:33m-l12-v2-fp16
bespoke-minicheck:7b
llama2:7b
all-minilm:22m-l6-v2-fp16
deepseek-r1:32b
llama3.3:70b
deepseek-r1:7b
deepseek-r1:14b
granite3.2-vision:2b-q4_K_M
llama3.2:3b-instruct-q5_K_M


In [ ]:
URL = "https://ollama-sam.inria.fr/"
login, passwd = "Bob","hiccup"

In [ ]:
# request = requests.get(URL + "api/tags"), auth = (login, passwd)

In [ ]:
# pour envoyer une requete prompt : 

prompt = "hi"
model = "mistral:7b"

request = requests.post("http://ollama.pl.sophia.inria.fr:8080/api/generate", data = {'prompt' :prompt , 'model' :model})

ConnectionError: HTTPConnectionPool(host='ollama.pl.sophia.inria.fr', port=8080): Max retries exceeded with url: /api/generate (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001FABAC4D7F0>: Failed to establish a new connection: [WinError 10061] Aucune connexion n’a pu être établie car l’ordinateur cible l’a expressément refusée'))

In [ ]:
answer = request.text

In [ ]:
answer

'{"error":"invalid character \'p\' looking for beginning of value"}'

In [21]:
!python -m spacy download en_core_web_md

     ---------------------------------------- 0.0/33.5 MB ? eta -:--:--
     --- ------------------------------------ 2.9/33.5 MB 16.8 MB/s eta 0:00:02
     ------- -------------------------------- 6.3/33.5 MB 15.5 MB/s eta 0:00:02
     ----------- --------------------------- 10.2/33.5 MB 16.4 MB/s eta 0:00:02
     --------------- ----------------------- 13.6/33.5 MB 16.5 MB/s eta 0:00:02
     --------------------- ----------------- 18.4/33.5 MB 17.5 MB/s eta 0:00:01
     --------------------------- ----------- 23.3/33.5 MB 18.7 MB/s eta 0:00:01
     -------------------------------- ------ 28.3/33.5 MB 19.5 MB/s eta 0:00:01
     ------------------------------------- - 32.5/33.5 MB 19.3 MB/s eta 0:00:01
     --------------------------------------- 33.5/33.5 MB 18.8 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_md')


In [22]:
# on essaie de faire des clusters sémantiques 
import spacy
from sklearn.cluster import KMeans
import numpy as np

nlp = spacy.load("en_core_web_md")

words = list(dico_filtré.keys())
vectors = np.array([nlp(w).vector for w in words])

kmeans = KMeans(n_clusters=7, random_state=0).fit(vectors)
clusters = {i: [] for i in range(7)}
for i, label in enumerate(kmeans.labels_):
    clusters[label].append(words[i])

print(clusters)


Exception in thread Thread-9 (_readerthread):
Traceback (most recent call last):
  File "c:\Users\ewenm\miniconda3\Lib\threading.py", line 1073, in _bootstrap_inner
    self.run()
  File "c:\Users\ewenm\miniconda3\Lib\site-packages\ipykernel\ipkernel.py", line 766, in run_closure
    _threading_Thread_run(self)
  File "c:\Users\ewenm\miniconda3\Lib\threading.py", line 1010, in run
    self._target(*self._args, **self._kwargs)
  File "c:\Users\ewenm\miniconda3\Lib\subprocess.py", line 1599, in _readerthread
    buffer.append(fh.read())
                  ^^^^^^^^^
  File "<frozen codecs>", line 322, in decode
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x82 in position 110: invalid start byte


{0: ['operation', 'process', 'application'], 1: ['study', 'model', 'emission', 'analysis', 'impact', 'method', 'approach', 'data', 'simulation', 'research', 'environmental', 'station', 'design', 'dynamic', 'framework', 'parameter', 'characteristic', 'alternative', 'modeling', 'objective', 'production', 'effective', 'smart'], 2: ['strategy', 'range', 'driving', 'adoption', 'conventional', 'behavior', 'bus'], 3: ['demand', 'reduce', 'development', 'transition', 'cycle', 'climate', 'cell'], 4: ['electric', 'system', 'power', 'transport', 'technology', 'performance', 'network', 'solution', 'consumption', 'optimization', 'efficiency', 'infrastructure', 'reduction', 'optimal', 'algorithm', 'capacity', 'storage', 'distribution', 'generation', 'mobility', 'supply', 'reducing', 'engine', 'operating', 'internal', 'efficient', 'mode', 'air', 'environment', 'integration', 'electrical'], 5: ['energy', 'electrification', 'fuel', 'electricity', 'carbon', 'grid', 'hybrid', 'renewable', 'source', 'diox

In [48]:
countries = df['COUNTRY'].sort_values().unique()
years = df['YEAR'].sort_values().unique()
countries = np.append(countries, ['ALL'])

In [49]:
keywords = np.array(list(dico_filtré.keys()))
keywords

array(['electric', 'vehicle', 'energy', 'system', 'study', 'model',
       'battery', 'charging', 'power', 'transport', 'electrification',
       'emission', 'cost', 'analysis', 'impact', 'demand', 'technology',
       'method', 'strategy', 'performance', 'approach', 'data', 'fuel',
       'simulation', 'electricity', 'carbon', 'reduce', 'network',
       'sector', 'transportation', 'research', 'solution', 'operation',
       'consumption', 'optimization', 'grid', 'efficiency',
       'infrastructure', 'development', 'environmental', 'management',
       'reduction', 'optimal', 'range', 'hybrid', 'renewable', 'policy',
       'driving', 'source', 'algorithm', 'station', 'capacity', 'design',
       'storage', 'distribution', 'process', 'dynamic', 'application',
       'generation', 'charge', 'economic', 'adoption', 'framework',
       'conventional', 'fleet', 'public', 'urban', 'dioxide',
       'greenhouse', 'sustainable', 'parameter', 'transition', 'behavior',
       'key', 'mobility

In [50]:
etiquettes_data = df.iloc[:, [2,6,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28]]  # à partir de la colonne 12, on prend tous les mots clés des études scientifiques 

In [51]:
def keywords_per_country(COUNTRY) : 
    etiquettes_series = pd.Series(etiquettes_data[etiquettes_data["COUNTRY"] == COUNTRY].iloc[:,2:].values.ravel())
    etiquettes_series = etiquettes_series.dropna().astype(str).str.strip().str.lower() 
    etiquettes_series = etiquettes_series[(etiquettes_series!='0')]
    keywords_unfiltered = list(etiquettes_series)
    return keywords_unfiltered

In [52]:
def keywords_per_year(YEAR) : 
    etiquettes_series = pd.Series(etiquettes_data[etiquettes_data["YEAR"] == YEAR].iloc[:,2:].values.ravel())
    etiquettes_series = etiquettes_series.dropna().astype(str).str.strip().str.lower() 
    etiquettes_series = etiquettes_series[(etiquettes_series!='0')]
    keywords_unfiltered = list(etiquettes_series)
    #keywords_filtered = [word for word in keywords_unfiltered if word in keywords]
    return keywords_unfiltered

In [53]:
def keywords_per_country_year(COUNTRY, YEAR) : 
    etiquettes_series = pd.Series(etiquettes_data[(etiquettes_data["YEAR"] == YEAR)&(etiquettes_data["COUNTRY"] == COUNTRY)].iloc[:,2:].values.ravel())
    etiquettes_series = etiquettes_series.dropna().astype(str).str.strip().str.lower() 
    etiquettes_series = etiquettes_series[(etiquettes_series!='0')]
    keywords_unfiltered = list(etiquettes_series)
    #keywords_filtered = [word for word in keywords_unfiltered if word in keywords]
    return list(set(keywords_unfiltered))

In [54]:
def keywords_split(keywords_unfiltered) :
    retour = []
    for keyword in keywords_unfiltered :
        retour += keyword.split()
    return retour

In [55]:
def filtered_keywords(unfiltered) : 
    return list(filter(lambda x: x in keywords, unfiltered))

In [56]:
def sans_doublons(liste) : 
    dico = {}
    retour = []
    for mot in liste : 
        if not mot in dico :
            dico[mot] = 0
            retour.append(mot)
    return retour

In [57]:
def country(COUNTRY) : 
    return sans_doublons(filtered_keywords(keywords_split(keywords_per_country(COUNTRY))))

def year(YEAR) : 
    return sans_doublons(filtered_keywords(keywords_split(keywords_per_year(YEAR))))

def country_and_year(COUNTRY, YEAR):
    return sans_doublons(filtered_keywords(keywords_split(keywords_per_country_year(COUNTRY, YEAR))))

In [ ]:
import dash
from dash import html, dcc, Input, Output,State
import plotly.graph_objects as go
import networkx as nx
import random

### On choisi le nombre de liens qu'on veut entre les mots 
liens= []

for a, b in itertools.combinations(df_cooc.columns, 2):
    if df_cooc.loc[a, b] > 500:  # on choisi de garder que les mots que reviennent ensemble de les titres plus de tant de fois.
        liens.append((a, b))

print(f"{len(liens)} liens trouvés avec cooccurrence > 1000")


# Couleurs des clusters de mots sémantiques
color_palette = ['red', 'orange', 'green', 'blue', 'purple', 'teal', 'brown']
node_color_map = {}
node_cluster_map = {}
for cluster_id, mots in clusters.items():
    for mot in mots:
        node_color_map[mot] = color_palette[cluster_id % len(color_palette)]
        node_cluster_map[mot] = cluster_id

##### GRAPHE 
#Cette partie permet de créer le graphe et d'ajouter des noeuds (nodes)  avec leurs tailles, et les liens (edges) entre les noeuds
G = nx.Graph()
for mot, poids in dico_filtré.items(): #on crée un noeud avec un mot clé de dico_filtré et sa taille
    G.add_node(mot)
    G.nodes[mot]['size'] = poids
G.add_edges_from(liens)

for node in G.nodes():
    if 'size' not in G.nodes[node]:
        G.nodes[node]['size'] = min(dico_filtré.values())
# Test pour vérifier que tous les noeuds ont une taille, sinon on lui attribue le min des poids (valeurs du dictionnaire)

# On commence par choisir des endroits distincts aléatoirement pour les différents foyers des clusters 
cluster_pos = {i: (random.uniform(-1, 1), random.uniform(-1, 1)) for i in clusters.keys()}
initial_pos = {mot: cluster_pos.get(node_cluster_map.get(mot, -1), (0, 0)) for mot in G.nodes()}

cluster_nodes = {i: [] for i in clusters.keys()}
for node in G.nodes():
    cluster = node_cluster_map.get(node, -1) # on récupère le cluster du noeud, ou -1 s'il n'est pas dans un cluster
    cluster_nodes[cluster].append(node) # on ajoute le noeud a la structure cluster correspondante

# On calcule les positions des noeuds manuellement
manual_pos = {}
for cluster_id, nodes in cluster_nodes.items():
    unconnected = [n for n in nodes if G.degree[n] == 0] # on récupère les noeuds non connectés
    angle_step = 360 / max(len(unconnected), 1) # on fait ceci pour éviter la division par zéro
    radius = 0.15 + len(unconnected) * 0.01  # rayon variable
    center_x, center_y = cluster_pos.get(cluster_id, (0, 0))
    for i, node in enumerate(unconnected): # on parcourt les noeuds non connectés
        angle_rad = (angle_step * i) * (3.1416 / 180)
        x = center_x + radius * random.uniform(0.9, 1.1) * np.cos(angle_rad)
        y = center_y + radius * random.uniform(0.9, 1.1) * np.sin(angle_rad)
        manual_pos[node] = (x, y)

#on prend les positions initiales et manuelles 
combined_pos = initial_pos.copy()
combined_pos.update(manual_pos)


pos = nx.spring_layout(G, pos=combined_pos, fixed=manual_pos.keys(), seed=42, k=0.4) # On utilise spring_layout pour ajuster les positions des noeuds en fonction des liens et des positions manuelles, avec une technique de minimisation de l'énergie, si on compare les distances à des coefficients de ressorts

app = dash.Dash(__name__)  #INTERFACE DASH
app.layout = html.Div([
    html.H1("Graphe sémantique par clusters"),      #Titre du graphe
    html.P("Select country:"),                      #Section pour choisir le pays
        dcc.Dropdown(
            id="dropdown",
            options= countries, # Liste des pays (cf ci dessus)
            value="ALBANIA",    # Valeur par défaut
            clearable=False,    # On ne peut pas effacer la valeur par défaut
        ),
    html.P("Select Year :"),    #Section pour choisir l'année
    dcc.Dropdown(
        id="year-dropdown", 
        options= years, # Liste des années (cf ci dessus)
        value="2020",  # Valeur par défaut
        clearable=False,
    ),
    dcc.Graph(id="graph")
])

@app.callback(
    Output("graph", "figure"),
    Input("graph", "hoverData"),
    State("graph", "relayoutData"),
    State("dropdown", "value"),
    State("year-dropdown", "value"),
) #On utilise un callback pour mettre à jour le graphe en fonction des interactions de l'utilisateur, hoverData pour le survol des noeuds, relayoutData pour les changements de zoom ou de position, et les valeurs des dropdowns pour le pays et l'année sélectionnés.

def update_graph(hoverData, relayoutData, country, year):
    if country == "ALL":
        keywords = keywords_per_year(year)
    else:
        keywords = country_and_year(country, year)  # On récupère les mots-clés en fonction du pays et de l'année sélectionnés
    
    # Reconstruire le node_trace
    node_x, node_y, node_text, node_size, node_colors = [], [], [], [], []
    max_poids = max(dico_filtré.values())
    for node in G.nodes():  # On parcourt tous les noeuds du graphe et on récupère les informations (positions, taille...)
        x, y = pos[node]
        node_x.append(x)
        node_y.append(y)
        node_text.append(node)
        node_size.append(G.nodes[node]['size'] / max_poids * 50 + 10)
        node_colors.append(node_color_map.get(node, "gray") if node in keywords else "#888") 
        # On utilise la couleur du cluster si le mot est dans les mots-clés, sinon on met une couleur grise (888)
    
    # On crée le tracé pour les noeuds du graphe
    node_trace = go.Scatter(
    x=node_x, y=node_y,
    mode="markers+text",
    text=node_text,
    hoverinfo="text",
    customdata=node_text,    #customdata est nécessaire ensuite pour ne pas renvoyer un message d'erreur quand on survole un noeud (ce qui est le cas si on utilise que text)
    marker=dict(
        color=node_colors,
        size=node_size,
        line_width=2
    ),
    textposition="top center" # on place le texte au dessus au milieu par rapport au noeud
    )

    fig = go.Figure(
            data=[node_trace],
            layout=go.Layout(
                showlegend=False,
                hovermode="closest",
                margin=dict(b=20, l=5, r=5, t=40),
                xaxis=dict(showgrid=False, zeroline=False),
                yaxis=dict(showgrid=False, zeroline=False),
            ),
        ) 
    #On récupère les données de la figure à tracer (le tracé des noeuds, les axes...)
    
    if relayoutData: #cette partie concerne le zoom et adapte l'échelle des deux axes lorsqu'on zoom
        if 'xaxis.range[0]' in relayoutData and 'xaxis.range[1]' in relayoutData: 
            fig.update_xaxes(range=[
                relayoutData['xaxis.range[0]'],
                relayoutData['xaxis.range[1]']
            ]) 
        if 'yaxis.range[0]' in relayoutData and 'yaxis.range[1]' in relayoutData:
            fig.update_yaxes(range=[
                relayoutData['yaxis.range[0]'],
                relayoutData['yaxis.range[1]']
            ])

    if hoverData is None: #si on ne survole pas de noeud, la figure reste identique 
        return fig 

    # Récupérer le nœud survolé
    try:
        node_id = hoverData["points"][0]["customdata"] # noeud survolé
    except (KeyError, IndexError, TypeError):
        # Fallback en cas de structure inattendue
        return dash.no_update

    # Arêtes liées à ce nœud
    filtered_edges = [e for e in G.edges() if node_id in e]

    # Générer uniquement les segments des arêtes concernées
    edge_x, edge_y = [], []
    for edge in filtered_edges:
        x0, y0 = pos[edge[0]]
        x1, y1 = pos[edge[1]]
        edge_x += [x0, x1, None] 
        edge_y += [y0, y1, None] 

    #Cette partie permet de tracer les arêtes reliées au noeud seleccioné
    edge_trace = go.Scatter(
        x=edge_x,
        y=edge_y,
        line=dict(width=2, color="#888"),
        hoverinfo="none",
        mode="lines",
    )
    
    fig.add_trace(edge_trace)   #la figure intègre le tracé des arêtes qui sont donc ajoutées et apparaissent sur le graphe

    return fig 
    
if __name__ == "__main__":
    app.run(debug=True)     #intructions pour faire tourner l'app

519 liens trouvés avec cooccurrence > 1000
